torch/csrc/api/src/optim/lbfgs.cpp

```diff
@@ -10,167 +10,602 @@
+      auto strong_wolfe_res = _strong_wolfe(obj_func, x_init, t, d, loss, flat_grad, gtd);
+      loss = std::get<0>(strong_wolfe_res);
+      flat_grad = std::get<1>(strong_wolfe_res);
+      t = std::get<2>(strong_wolfe_res);
+      ls_func_evals = std::get<3>(strong_wolfe_res);
+      _add_grad(t, d);
+      opt_cond = (val(flat_grad.abs().max()) <= tolerance_grad);
+    } else {
+      // no line search, simply move with fixed-step
+      _add_grad(t, d);
+      if (n_iter != max_iter) {
+        // re-evaluate function only if not in last iteration
+        // the reason we do this: in a stochastic setting,
+        // no use to re-evaluate that function here
+        {
+          torch::AutoGradMode enable_grad(true);
+          loss = closure();
+        }
+        flat_grad = _gather_flat_grad();
+        opt_cond = val(torch::max(flat_grad.abs())) <= tolerance_grad;
+        ls_func_evals = 1;
+      }
     }
+    // update func eval
+    current_evals += ls_func_evals;
+    state.func_evals(state.func_evals() + ls_func_evals);
+
+    // ############################################################
+    // # check conditions
+    // ############################################################
+    if (n_iter == max_iter) break;
+
+    if (current_evals >= *max_eval) break;
+
+    // optimal condition
+    if (opt_cond) break;
+
+    // lack of progress
+    if (val(d.mul(t).abs().max()) <= tolerance_change) break;
+
+    if (val(abs(loss - prev_loss)) < tolerance_change) break;
   }
+
   return orig_loss;
 }
 
+void LBFGS::add_parameters(const std::vector<Tensor>& parameters) {
+  return _add_parameters_new_design(parameters);
+}
+
+const std::vector<Tensor>& LBFGS::parameters() const noexcept {
+  return _parameters_new_design();
+}
+
+std::vector<Tensor>& LBFGS::parameters() noexcept {
+  return _parameters_new_design();
+}
+
+size_t LBFGS::size() const noexcept {
+  return _size_new_design();
+}
+
 void LBFGS::save(serialize::OutputArchive& archive) const {
   serialize(*this, archive);
 }
 
 void LBFGS::load(serialize::InputArchive& archive) {
-  serialize(*this, archive);
+  IValue pytorch_version;
+  if (archive.try_read("pytorch_version", pytorch_version)) {
+    serialize(*this, archive);
+  }
+  else { // deserializing archives saved in old format (prior to version 1.5.0)
+    TORCH_WARN(
+      "Your serialized LBFGS optimizer is still using the old serialization format. "
+      "The func_evals and n_iter value in state will be set to 0 because the old RMSprop optimizer didn't save these values."
+      "You should re-save your LBFGS optimizer to use the new serialization format.");
+    Tensor d, t, H_diag, prev_flat_grad, prev_loss;
+    std::deque<Tensor> old_dirs, old_stps;
+    archive("d", d);
+    archive("t", t);
+    archive("H_diag", H_diag);
+    archive("prev_flat_grad", prev_flat_grad);
+    archive("prev_loss", prev_loss);
```

Comment:
Should we pass `/*is_buffer=*/true` for the above five calls? (The old C++ version does so)